app/src/main/java/com/zulip/android/activities/LoginActivity.java

```diff
@@ -265,6 +182,93 @@ public void onTaskFailure(String result) {
         }
     }
 
+    private void setupSignIN(String clientId) {
+        AuthorizationServiceConfiguration serviceConfiguration = new AuthorizationServiceConfiguration(
+                Uri.parse("https://accounts.google.com/o/oauth2/v2/auth") /* auth endpoint */,
+                Uri.parse("https://www.googleapis.com/oauth2/v4/token") /* token endpoint */
+        );
+
+        Uri redirectUri = Uri.parse("com.zulip.android:/oauth2callback");
+        AuthorizationRequest.Builder builder = new AuthorizationRequest.Builder(
+                serviceConfiguration,
+                clientId,
+                AuthorizationRequest.RESPONSE_TYPE_CODE,
+                redirectUri
+        );
+        builder.setScopes("profile", "email");
+        AuthorizationRequest request = builder.build();
+        AuthorizationService authorizationService = new AuthorizationService(LoginActivity.this);
+        String action = "com.zulip.android.HANDLE_AUTHORIZATION_RESPONSE";
+        Intent postAuthorizationIntent = new Intent(action);
+        PendingIntent pendingIntent = PendingIntent.getActivity(LoginActivity.this, request.hashCode(), postAuthorizationIntent, 0);
+        authorizationService.performAuthorizationRequest(request, pendingIntent);
+    }
+
+    private void checkIntent(@Nullable Intent intent) {
+        if (intent != null && intent.getAction() != null) {
+            String action = intent.getAction();
+            switch (action) {
+                case "com.zulip.android.HANDLE_AUTHORIZATION_RESPONSE":
+                    if (!intent.hasExtra(USED_INTENT)) {
+                        handleAuthorizationResponse(intent);
+                        intent.putExtra(USED_INTENT, true);
+                    }
+                    break;
+                default:
+                    // do nothing
+            }
+        }
+    }
+
+    private void handleAuthorizationResponse(Intent intent) {
```

Comment:
This method is hard to read. If there's a way to clean it up or break out more of the logic into smaller sub-methods that'd be 👍